parallel-exec/lib/lib.go

```diff
@@ -0,0 +1,418 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package lib
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"log"
+	"os"
+	"os/exec"
+	"os/signal"
+	"runtime"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+)
+
+const (
+	// EventTypeStarted says that the runner started.
+	EventTypeStarted EventType = iota
+	// EventTypeCmdStarted says that a command started.
+	EventTypeCmdStarted
+	// EventTypeCmdFinished says that a command finished.
+	EventTypeCmdFinished
+	// EventTypeFinished says that the runner finished.
+	EventTypeFinished
+
+	defaultFastFail = false
+)
+
+var (
+	defaultMaxConcurrentCmds = runtime.NumCPU()
+	defaultEventHandler      = func(event Event) {
+		data, err := json.Marshal(event)
+		if err != nil {
+			log.Print(event.Type())
+		}
+		log.Print(string(data))
+	}
+	defaultTimer = time.Now
+
+	errCmdFailed   = errors.New("command failed")
+	errInterrupted = errors.New("interrupted")
+)
+
+// EventType is an event type during the runner's run call.
+type EventType int
+
+// String returns a string representation of the EventType.
+func (e EventType) String() string {
+	switch e {
+	case EventTypeStarted:
+		return "started"
+	case EventTypeCmdStarted:
+		return "cmd_started"
+	case EventTypeCmdFinished:
+		return "cmd_finished"
+	case EventTypeFinished:
+		return "finished"
+	default:
+		return strconv.Itoa(int(e))
+	}
+}
+
+// ParseEventType parses an EventType from it's string representation.
+func ParseEventType(eventTypeString string) (EventType, error) {
+	switch eventTypeString {
+	case "started":
+		return EventTypeStarted, nil
+	case "cmd_started":
+		return EventTypeCmdStarted, nil
+	case "cmd_finished":
+		return EventTypeCmdFinished, nil
+	case "finished":
+		return EventTypeFinished, nil
+	default:
+		return 0, fmt.Errorf("unknown EventType string: %s", eventTypeString)
+	}
+}
+
+// Event is an event that happens during the runner's run call.
+type Event interface {
+	Type() EventType
+	Time() time.Time
+	Fields() map[string]string
+}
+
+// EventHandler handles events.
+type EventHandler func(Event)
+
+// Timer returns the current time.
+type Timer func() time.Time
```

Comment:
I don't think `Timer` is the right word for a function that returns the current time. Maybe `Clock`? or just use `func() time.Time` directly